File: mutate4go/workspace_test.go
package mutate4go

import (
	"os"
	"path/filepath"
	"testing"
)

func TestPrepareWorkerRootsCreatesWorkerCopiesWithoutCopyingMutateOutput(t *testing.T) {
	root := t.TempDir()
	writeFile(t, filepath.Join(root, "go.mod"), "module example.com/demo\n\ngo 1.18\n")
	writeFile(t, filepath.Join(root, "demo", "app.go"), "package demo\n")
	writeFile(t, filepath.Join(root, ".mutate", "workers", "old", "ignored.txt"), "ignored")

	workerRoots, cleanup, err := prepareWorkerRoots(root, 2)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	defer cleanup()

	workerRoot := workerRoots[0]
	if _, err := os.Stat(filepath.Join(workerRoot, "go.mod")); err != nil {
		t.Fatalf("expected go.mod in worker root: %v", err)
	}
	if _, err := os.Stat(filepath.Join(workerRoot, "demo", "app.go")); err != nil {
		t.Fatalf("expected source file in worker root: %v", err)
	}
	if _, err := os.Stat(filepath.Join(workerRoot, ".mutate", "workers", "old", "ignored.txt")); !os.IsNotExist(err) {
		t.Fatalf("expected mutate output to be excluded, got err=%v", err)
	}
}

func TestPrepareWorkerRootsCleanupRemovesRunDirectory(t *testing.T) {
	root := t.TempDir()
	writeFile(t, filepath.Join(root, "go.mod"), "module example.com/demo\n\ngo 1.18\n")
	writeFile(t, filepath.Join(root, "demo", "app.go"), "package demo\n")

	workerRoots, cleanup, err := prepareWorkerRoots(root, 1)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	runRoot := filepath.Dir(workerRoots[0])
	writeFile(t, filepath.Join(workerRoots[0], ".mutate", "tmp", "result.txt"), "done")

	cleanup()

	if _, err := os.Stat(runRoot); !os.IsNotExist(err) {
		t.Fatalf("expected run root removed, got err=%v", err)
	}
}

File: mutate4rs/tests/workspace_test.rs
use std::{
    fs,
    path::PathBuf,
    time::{SystemTime, UNIX_EPOCH},
};

use mutate4rs::prepare_worker_roots;

#[test]
fn prepare_worker_roots_copies_project_without_mutate_output() {
    let root = temp_dir();
    fs::create_dir_all(root.join("src")).unwrap();
    fs::write(root.join("src/lib.rs"), "pub fn value() -> bool { true }\n").unwrap();
    fs::create_dir_all(root.join(".mutate/workers/old")).unwrap();
    fs::write(root.join(".mutate/workers/old/ignored.txt"), "ignored").unwrap();

    let workspaces = prepare_worker_roots(&root, 2).unwrap();

    let worker_root = PathBuf::from(&workspaces.worker_roots[0]);
    assert!(worker_root.join("src/lib.rs").exists());
    assert!(!worker_root.join(".mutate/workers/old/ignored.txt").exists());
    workspaces.close().unwrap();
}

#[test]
fn close_removes_run_root() {
    let root = temp_dir();
    fs::create_dir_all(root.join("src")).unwrap();
    fs::write(root.join("src/lib.rs"), "pub fn value() -> bool { true }\n").unwrap();

    let workspaces = prepare_worker_roots(&root, 1).unwrap();
    let run_root = PathBuf::from(&workspaces.run_root);
    fs::create_dir_all(run_root.join("worker-1/.mutate/tmp")).unwrap();
    fs::write(run_root.join("worker-1/.mutate/tmp/result.txt"), "done").unwrap();

    workspaces.close().unwrap();
    assert!(!run_root.exists());
}

fn temp_dir() -> PathBuf {
    let unique = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    let dir = std::env::temp_dir().join(format!("mutate4rs-workspace-{unique}"));
    fs::create_dir_all(&dir).unwrap();
    dir
}
